### ISSUE TITLE
Prefixes for dot files are not bijective

### ISSUE THREAD

**Issue by jhamfler** [twpayne/chezmoi#438]
#### Description of issue

When adding a file named `dot_<something>` it is added to the repo as such. When applying it a file `.<something>` is created instead. This is an issue with prefixes in general.

#### Command that failed

```
jh@d10:~$ echo "test" >> dot_testfile
jh@d10:~$ ls -la
-rw-r--r--  1 jh jh    5 Sep 29 17:11 dot_testfile
jh@d10:~$ chezmoi add dot_testfile
jh@d10:~$ chezmoi cd
jh@d10:~/.local/share/chezmoi$ ls -la
-rw-r--r--  1 jh jh    5 Sep 29 17:11 dot_testfile
drwxr-xr-x  8 jh jh 4096 Sep 29 14:18 .git
jh@d10:~/.local/share/chezmoi$ exit
jh@d10:~$ chezmoi apply
jh@d10:~$ ls -la
-rw-r--r--  1 jh jh    5 Sep 29 17:11 dot_testfile
-rw-r--r--  1 jh jh    5 Sep 29 17:12 .testfile
```

#### Output of `chezmoi --version`

```
chezmoi version 1.6.5, commit 3733485a2e018dfe174423c5c58edd6d03f0e3c6, built at 2019-09-20T12:17:46Z
```

**Comment by twpayne** [comment]
Yes, you are correct. Does this cause a practical problem for you?

**Comment by jhamfler** [comment]
You got me. It doesn't cause a practical problem, but it might sometime. Better safe than sorry in my opinion. I can also imagine adding a file with a long filename where the additional prefix would be over the limit of the filesystem (e.g. 255 bytes for ext4).

**Comment by twpayne** [comment]
There are lots of potential problems with filenames: different systems allow different characters, case sensitivity, whether case is preserved, and I'm sure that there are loads more when you add Windows to the mix.

I think (but have not verified) that you can fix the non-bijectivity by including something like a `literal_` prefix, where everything after the first occurrence of `literal_` is taken as the rest of the filename, so `dot_testfile` would be named `literal_dot_testfile` in the source state.

The `.tmpl` extension is not so much of a problem as you can repeat it twice and only the last one will be stripped (although the contents will be treated as a template which may cause other problems).

However, in practice, dotfiles tend to have short, simple names, so I propose to only fix this once it is reported as causing a genuine problem.

I'll add an entry to the FAQ about this.

**Comment by jhamfler** [comment]
Seems fair.

**Comment by twpayne** [comment]
> However, in practice, dotfiles tend to have short, simple names, so I propose to only fix this once it is reported as causing a genuine problem.

And it finally did! The culprit was a file called [`run_tester.lua`](https://github.com/nvim-lua/plenary.nvim/blob/master/scratch/run_tester.lua). #1283 contains the fix.

Note that I'll mark this as fixed although the mapping between the target filenames and source filenames remains non bijective.  Multiple source filenames (e.g. `name`, `literal_name`, `name.literal`, and `literal_name.literal`) map to the same target filename (`name`). However, each target filename has a canonical source filename (`literal_` and `.literal` are only used if needed). As the source filenames are under chezmoi's control, and the set of source filenames is strictly larger than the set of target filenames (and therefore cardinality cannot match and therefore a bijection is impossible), I'll mark this as fixed once #1283 is merged.

Thanks again for the discussion here - really appreciated.
